bindings/python/module.cpp

```diff
@@ -72,4 +77,36 @@ PYBIND11_MODULE(_pyghex, m)
     pyghex::unstructured::register_field_descriptor(m);
     pyghex::unstructured::register_pattern(m);
     pyghex::unstructured::register_communication_object(m);
+
+    m.def("expose_cpp_ptr", [](pyghex::context_shim* obj) {return reinterpret_cast<uintptr_t>(&obj->m);} );
+    
+    gridtools::for_each<
+        gridtools::meta::transform<gridtools::meta::list, pyghex::unstructured::communication_object_specializations>>(
+        [&m](auto l)
+        {
+            using type = gridtools::meta::first<decltype(l)>;
+            m.def("expose_cpp_ptr", [](type* obj) {return reinterpret_cast<uintptr_t>(obj);} );
+        });
+
+    gridtools::for_each<
+        gridtools::meta::transform<gridtools::meta::list, pyghex::unstructured::make_pattern_traits_specializations>>(
+        [&m](auto l)
+        {
+            using type = gridtools::meta::first<decltype(l)>;
+            using halo_gen = typename type::halo_gen;
+            using domain_range = typename type::domain_range;
+            using grid_type = ghex::unstructured::grid;
+            using pattern_container =
+                decltype(ghex::make_pattern<grid_type>(std::declval<ghex::context&>(),
+                    std::declval<halo_gen&>(), std::declval<domain_range&>()));
+            m.def("expose_cpp_ptr", [](pattern_container* obj) {return reinterpret_cast<uintptr_t>(obj);} );
+        });
+
+    gridtools::for_each<
+        gridtools::meta::transform<gridtools::meta::list, pyghex::unstructured::domain_descriptor_specializations>>(
+        [&m](auto l)
+        {
+            using type = gridtools::meta::first<decltype(l)>;
+            m.def("expose_cpp_ptr", [](type* obj) {return reinterpret_cast<uintptr_t>(obj);} );
+        });
```

Comment:
I would move those functions into their respective source files to have only one loop over the types.